### PR TITLE
feat(streaming): vnode bitmap updates for state table

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

When scaling, we'll update the vnode bitmap for the state tables in old executors. This is mainly for assertion purposes since there seems to be no executor accessing the storage across different vnodes, but still necessary for our debugging.

The logic for updating should be per-executor. This PR implements it for:
- hash join
- hash agg
- dynamic filter
- group top-n

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- #4645

### CODE REVIEW

File: src/stream/src/from_proto/hash_agg.rs [hunk 1]
```diff
@@ -45,6 +47,7 @@ impl<S: StateStore> HashKeyDispatcher for HashAggExecutorDispatcher<S> {
             args.input,
             args.agg_calls,
             args.pk_indices,
+            args.actor_id,
```
Comment (fuyufjh) [inline src/stream/src/from_proto/hash_agg.rs:50]: May introduce an executor context to store actor_id as well as other common info shared by all executors.

File: src/stream/src/executor/top_n_executor.rs [hunk 2]
```diff
@@ -48,6 +49,10 @@ pub trait TopNExecutorBase: Send + 'static {
     /// See [`Executor::identity`].
     fn identity(&self) -> &str;
 
+    /// Update the vnode bitmap for the state tables, only used by Group Top-N since it's
+    /// distributed.
+    fn update_state_table_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) {}
```
Comment (fuyufjh) [inline src/stream/src/executor/top_n_executor.rs:54]: Since only group TopN uses this, why not just impl for group_top_n?

**Review by fuyufjh — APPROVED** [review]
Generally LGTM

File: src/stream/src/from_proto/hash_agg.rs (hunk 1 above)
Comment (BugenZhao) [reply src/stream/src/from_proto/hash_agg.rs:50]: Count on @xxchan 's `Context`. 😄

File: src/stream/src/executor/top_n_executor.rs (hunk 2 above)
Comment (BugenZhao) [reply src/stream/src/executor/top_n_executor.rs:54]: The `execute` is actually implemented by `TopNExecutorWrapper` and there's no way to know which inner `base` it is. So we need a generic interface for this. 🤔

**Review by skyzh — APPROVED** [review]
LGTM. Just one question: why we choose to update vnode bitmaps instead of creating new executors?

**Comment by BugenZhao** [comment]
> LGTM. Just one question: why we choose to update vnode bitmaps instead of creating new executors?

For scaling out, we create executors for new parallelisms and keep the old ones, whose vnode bitmaps need to be updated. By keeping the old executors, the procedure of scaling can be much easier, and also the operator cache can be reserved.

**Pull request merged**